File: install/bootstrap.py
#!/usr/bin/env python3
"""Bootstrap a SELF-CONTAINED, repo-level Genesis workspace at <target_repo>/.genesis/.

Genesis's expertise store, memory, hooks, and team live PER REPOSITORY (never global). When Sensei is asked
to build an agent in a repo that has no `.genesis/` workspace yet, this creates one in a single deterministic
call, so the repo becomes a standalone Genesis: its own store + hooks + memory DB + base modules + sensei/method.

What it does (idempotent, cross-platform):
  1. Copies the functional Genesis tree into <repo>/.genesis/: expertise/ (+manifests/, required.json),
     hooks/, skills/ (the team skills build-agent/research-expertise + the spec-forge suite), team/
     (sensei+method personas/behaviors), install/ (assemble/install/bootstrap), and bin/ if present.
  2. Copies ONLY the memory-server binary + ONNX model (not the multi-GB Rust build tree) into
     <repo>/.genesis/server/ — so the workspace is fully self-contained (no dependency on the central Genesis).
  3. Registers a repo-local `genesis-memory` MCP server in <repo>/.mcp.json (merged, not clobbered), pointing
     the binary + model + SQLite DB at <repo>/.genesis/ paths.
  4. Installs sensei + method into <repo>/.claude/agents/ via the assembler with genesis_home = <repo>/.genesis,
     so their hook commands reference <repo>/.genesis/hooks/*.py and the validate/review hooks resolve the
     store at <repo>/.genesis/expertise/ through their existing HOOK_DIR/../expertise relative logic.
Re-running refreshes the copied tree + agents WITHOUT clobbering an existing memory DB.

usage: bootstrap.py <target_repo> [genesis_home]     # genesis_home from $GENESIS_HOME, else this repo root
"""
import json, os, shutil, subprocess, sys

TEAM = ["sensei", "method"]


def _ignore_junk(_dir, names):
    return {n for n in names if n == "__pycache__" or n.endswith(".pyc")}


def _copy_tree(src, dst):
    shutil.copytree(src, dst, dirs_exist_ok=True, ignore=_ignore_junk)


# The managed .gitignore block. COMMIT the agent brain (expertise + hooks) and portable memory
# (.genesis/memory/*.jsonl); IGNORE machine-local junk (the DB, server binary+model, caches, logs,
# the absolute-path .mcp.json). Rewritten in place between the sentinels on every bootstrap.
GITIGNORE_START = "# >>> genesis runtime (managed by bootstrap) >>>"
GITIGNORE_END = "# <<< genesis runtime <<<"
GITIGNORE_BLOCK = "\n".join([
    GITIGNORE_START,
    "# Commit the agent brain (expertise + hooks) and portable memory (JSONL) so agents and their",
    "# learned memory travel with the repo across systems. Ignore machine-local / regenerable junk.",
    ".genesis/*",
    "!.genesis/expertise/",
    "!.genesis/hooks/",
    "!.genesis/memory/",
    ".genesis/**/__pycache__/",
    ".genesis/expertise/.genesis/",
    ".genesis/memory/*.tmp",
    "*.db",
    ".mcp.json",
    GITIGNORE_END,
    "",
])


def _merge_gitignore(target):
    """Idempotently write the managed genesis block into <target>/.gitignore.

    Replaces an existing managed block (between sentinels) in place; otherwise appends. Never
    touches the user's own lines. Returns a note if a conflicting blanket `.genesis/` ignore is
    found outside the managed block (it would defeat the re-includes).
    """
    path = os.path.join(target, ".gitignore")
    try:
        existing = open(path, encoding="utf-8").read()
    except Exception:
        existing = ""
    if GITIGNORE_START in existing and GITIGNORE_END in existing:
        pre = existing.split(GITIGNORE_START)[0]
        post = existing.split(GITIGNORE_END, 1)[1]
        merged = pre.rstrip("\n") + ("\n\n" if pre.strip() else "") + GITIGNORE_BLOCK + post.lstrip("\n")
    else:
        sep = "" if (not existing or existing.endswith("\n")) else "\n"
        merged = existing + sep + ("\n" if existing.strip() else "") + GITIGNORE_BLOCK
    with open(path, "w", encoding="utf-8", newline="\n") as f:
        f.write(merged)
    # Warn if a blanket `.genesis/` ignore lives OUTSIDE our block — it would win and defeat the re-includes.
    outside = merged.split(GITIGNORE_START)[0] + merged.split(GITIGNORE_END, 1)[-1]
    conflict = any(ln.strip() in (".genesis", ".genesis/") for ln in outside.splitlines())
    return "conflicting blanket '.genesis/' ignore present — remove it" if conflict else None


def main():
    if len(sys.argv) < 2:
        sys.exit("usage: bootstrap.py <target_repo> [genesis_home]  (or set $GENESIS_HOME)")
    target = os.path.abspath(sys.argv[1])
    gh = os.path.abspath(sys.argv[2]) if len(sys.argv) > 2 \
        else os.path.abspath(os.environ["GENESIS_HOME"]) if os.environ.get("GENESIS_HOME") \
        else os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    if not os.path.isdir(target):
        sys.exit(f"target repo not found: {target}")
    if os.path.abspath(gh) == os.path.abspath(os.path.join(target, ".genesis")):
        sys.exit("refusing to bootstrap a genesis into itself.")

    dest = os.path.join(target, ".genesis")

    # 1. Copy the functional Genesis tree (self-contained). `skills/` now holds the team skills (build-agent,
    #    research-expertise) + the spec-forge suite, so it travels too — the assembler wires sensei's skills
    #    from <dest>/skills/ (built-ins) via install_named_skills.
    for sub in ("expertise", "hooks", "skills", "team", "install"):
        s = os.path.join(gh, sub)
        if not os.path.isdir(s):
            sys.exit(f"genesis home is missing {sub}/ at {gh} — is {gh} a Genesis install?")
        _copy_tree(s, os.path.join(dest, sub))

    # 1b. Best-effort: carry the memory-server LAUNCHER (bin/) into <repo>/.genesis/bin/ if it exists (built
    #     by the plugin's binary-delivery worker). The compiled binary is also copied below (step 2), so a
    #     bootstrapped repo is self-contained regardless of whether the launcher is present yet.
    bin_src = os.path.join(gh, "bin")
    if os.path.isdir(bin_src):
        _copy_tree(bin_src, os.path.join(dest, "bin"))

    # 2. Copy ONLY the server binary + model (NOT server/target/, which is multi-GB build output).
    exe = ".exe" if os.name == "nt" else ""
    bin_src = os.path.join(gh, "server", "target", "release", "genesis-memory-server" + exe)
    if not os.path.exists(bin_src) and os.path.exists(bin_src + ".exe"):
        bin_src, exe = bin_src + ".exe", ".exe"
    if not os.path.exists(bin_src):
        sys.exit(f"memory-server binary not found at {bin_src} — build it first (cd server && cargo build --release).")
    bin_dst_dir = os.path.join(dest, "server", "target", "release")
    os.makedirs(bin_dst_dir, exist_ok=True)
    bin_dst = os.path.join(bin_dst_dir, "genesis-memory-server" + exe)
    shutil.copy2(bin_src, bin_dst)
    try:
        os.chmod(bin_dst, 0o755)
    except Exception:
        pass
    models_src = os.path.join(gh, "server", "models")
    if not os.path.isdir(models_src):
        sys.exit(f"model dir not found at {models_src}.")
    _copy_tree(models_src, os.path.join(dest, "server", "models"))

    # 3. Memory DB lives under .genesis/ (server creates the file on first use). Never clobber an existing one.
    #    The DB is machine-local + regenerable; the PORTABLE, COMMITTED form of memory is the JSONL export
    #    at .genesis/memory/memory.jsonl (see step 4b + the .gitignore in step 6). On a fresh clone the DB is
    #    absent but the JSONL is present, so the server rebuilds (re-embeds) the memory on first run.
    mem_db = os.path.join(dest, "memory.db")
    mem_export = os.path.join(dest, "memory", "memory.jsonl")

    # 4. Register the repo-local memory MCP server in <repo>/.mcp.json (merge; preserve other servers).
    mcp_path = os.path.join(target, ".mcp.json")
    try:
        mcp = json.load(open(mcp_path, encoding="utf-8"))
    except Exception:
        mcp = {}
    if not isinstance(mcp, dict):
        mcp = {}
    mcp.setdefault("mcpServers", {})
    mcp["mcpServers"]["genesis-memory"] = {
        "command": bin_dst,
        "args": [],
        "env": {"GENESIS_MODEL_DIR": os.path.join(dest, "server", "models"),
                "GENESIS_MEMORY_DB": mem_db,
                # The committed, cross-system-portable mirror. The server snapshots to this after every
                # store/consolidate and rebuilds from it when the DB is empty (fresh clone).
                "GENESIS_MEMORY_EXPORT": mem_export},
    }
    with open(mcp_path, "w", encoding="utf-8") as f:
        json.dump(mcp, f, indent=2)
        f.write("\n")

    # 6. Ensure the repo's .gitignore commits the agent BRAIN (expertise + hooks) and PORTABLE MEMORY
    #    (.genesis/memory/*.jsonl) while ignoring machine-local junk (the DB, the server binary+model,
    #    caches, logs, the absolute-path .mcp.json). Without this, agents' memory + corrections never
    #    travel across systems — the flaw this block fixes.
    gitignore_note = _merge_gitignore(target)

    # 5. Install sensei + method into <repo>/.claude/agents/, wired to the REPO-LEVEL .genesis/ home.
    assembler = os.path.join(dest, "install", "assemble.py")
    installed = []
    for name in TEAM:
        src = os.path.join(dest, "team", name)
        r = subprocess.run([sys.executable, assembler, src, name, target, dest],
                           capture_output=True, text=True)
        if r.returncode != 0:
            sys.exit(f"assemble {name} failed: {r.stderr.strip() or r.stdout.strip()}")
        installed.append(json.loads(r.stdout.strip())["agent"])

    print(json.dumps({
        "bootstrapped": dest,
        "target_repo": target,
        "agents_installed": installed,
        "agents_dir": os.path.join(target, ".claude", "agents"),
        "memory_server": bin_dst,
        "model_dir": os.path.join(dest, "server", "models"),
        "memory_db": mem_db,
        "memory_db_preexisting": os.path.exists(mem_db),
        "memory_export": mem_export,
        "memory_export_preexisting": os.path.exists(mem_export),
        "gitignore": os.path.join(target, ".gitignore"),
        "gitignore_warning": gitignore_note,
        "mcp_json": mcp_path,
        "next": "Open Claude Code in the repo; talk to Sensei to build agents (expertise via the research-expertise skill).",
    }, indent=2))


if __name__ == "__main__":
    main()

File: install/bootstrap.js
#!/usr/bin/env node
/* Bootstrap a SELF-CONTAINED, repo-level Genesis workspace at <target_repo>/.genesis/.

   Faithful Node (CommonJS, stdlib-only) port of bootstrap.py, updated for the GitHub-Releases distribution:
   the generated repo `.mcp.json` launches the server via a small Node launcher (`.genesis/bin/genesis-memory.js`)
   that downloads + SHA256-verifies the platform binary + model from the GitHub Release and caches them — so
   there is NO local Rust build and NO committed multi-GB binary/model. Node is the only runtime a user needs.

   Genesis's expertise store, memory, hooks, and team live PER REPOSITORY (never global). When Sensei is asked
   to build an agent in a repo that has no `.genesis/` workspace yet, this creates one in a single deterministic
   call, so the repo becomes a standalone Genesis: its own store + hooks + memory DB + base modules + sensei/method.

   What it does (idempotent, cross-platform):
     1. Copies the functional Genesis tree into <repo>/.genesis/: expertise/ (+manifests/, required.json),
        hooks/ (hooks.json + the run.js resolver), skills/ (team skills + the spec-forge suite), team/,
        install/, and bin/ (the launcher). Then stages the native genesis-hook binary into <repo>/.genesis/bin.
     2. Registers a repo-local `genesis-memory` MCP server in <repo>/.mcp.json (merged, not clobbered) that runs
        `node <repo>/.genesis/bin/genesis-memory.js`, pointing the memory DB + portable JSONL export at <repo>/.genesis/.
     3. Installs sensei + method into <repo>/.claude/agents/ via the assembler with genesis_home = <repo>/.genesis,
        so their hook commands invoke <repo>/.genesis/bin/genesis-hook directly and pass
        --expertise <repo>/.genesis/expertise.
   Re-running refreshes the copied tree + agents WITHOUT clobbering an existing memory DB.

   usage: bootstrap.js <target_repo> [genesis_home]     # genesis_home from $GENESIS_HOME, else this repo root
*/
"use strict";
const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

const TEAM = ["sensei", "method"];

function isDir(p) {
  try {
    return fs.statSync(p).isDirectory();
  } catch (e) {
    return false;
  }
}

function fail(msg) {
  process.stderr.write(msg + "\n");
  process.exit(1);
}

function copyTree(src, dst) {
  // shutil.copytree(dirs_exist_ok=True) equivalent, skipping __pycache__ dirs and *.pyc files.
  fs.cpSync(src, dst, {
    recursive: true,
    filter: (s) => path.basename(s) !== "__pycache__" && !s.endsWith(".pyc"),
  });
}

// The managed .gitignore block. COMMIT the agent brain (expertise + hooks) and portable memory
// (.genesis/memory/*.jsonl); IGNORE machine-local junk (the DB, caches, logs, the absolute-path .mcp.json).
// Rewritten in place between the sentinels on every bootstrap.
const GITIGNORE_START = "# >>> genesis runtime (managed by bootstrap) >>>";
const GITIGNORE_END = "# <<< genesis runtime <<<";
const GITIGNORE_BLOCK = [
  GITIGNORE_START,
  "# Commit the agent brain (expertise + hooks) and portable memory (JSONL) so agents and their",
  "# learned memory travel with the repo across systems. Ignore machine-local / regenerable junk.",
  ".genesis/*",
  "!.genesis/expertise/",
  "!.genesis/hooks/",
  "!.genesis/memory/",
  ".genesis/**/__pycache__/",
  ".genesis/expertise/.genesis/",
  ".genesis/memory/*.tmp",
  "*.db",
  ".mcp.json",
  GITIGNORE_END,
  "",
].join("\n");

function mergeGitignore(target) {
  // Idempotently write the managed genesis block into <target>/.gitignore.
  //
  // Replaces an existing managed block (between sentinels) in place; otherwise appends. Never touches the
  // user's own lines. Returns a note if a conflicting blanket `.genesis/` ignore is found outside the managed
  // block (it would defeat the re-includes).
  const p = path.join(target, ".gitignore");
  let existing;
  try {
    existing = fs.readFileSync(p, { encoding: "utf-8" });
  } catch (e) {
    existing = "";
  }
  let merged;
  if (existing.includes(GITIGNORE_START) && existing.includes(GITIGNORE_END)) {
    const pre = existing.slice(0, existing.indexOf(GITIGNORE_START));
    const post = existing.slice(existing.indexOf(GITIGNORE_END) + GITIGNORE_END.length);
    merged = pre.replace(/\n+$/, "") + (pre.trim() ? "\n\n" : "") + GITIGNORE_BLOCK + post.replace(/^\n+/, "");
  } else {
    const sep = !existing || existing.endsWith("\n") ? "" : "\n";
    merged = existing + sep + (existing.trim() ? "\n" : "") + GITIGNORE_BLOCK;
  }
  fs.writeFileSync(p, merged, { encoding: "utf-8" });
  // Warn if a blanket `.genesis/` ignore lives OUTSIDE our block — it would win and defeat the re-includes.
  const outside =
    merged.slice(0, merged.indexOf(GITIGNORE_START)) +
    merged.slice(merged.indexOf(GITIGNORE_END) + GITIGNORE_END.length);
  const conflict = outside
    .split(/\r\n|\r|\n/)
    .some((ln) => ln.trim() === ".genesis" || ln.trim() === ".genesis/");
  return conflict ? "conflicting blanket '.genesis/' ignore present — remove it" : null;
}

function main() {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    fail("usage: bootstrap.js <target_repo> [genesis_home]  (or set $GENESIS_HOME)");
  }
  const target = path.resolve(args[0]);
  const gh =
    args.length > 1
      ? path.resolve(args[1])
      : process.env.GENESIS_HOME
      ? path.resolve(process.env.GENESIS_HOME)
      : path.dirname(__dirname);
  if (!isDir(target)) {
    fail(`target repo not found: ${target}`);
  }
  if (path.resolve(gh) === path.resolve(path.join(target, ".genesis"))) {
    fail("refusing to bootstrap a genesis into itself.");
  }

  const dest = path.join(target, ".genesis");

  // 1. Copy the functional Genesis tree (self-contained). `skills/` holds the team skills (build-agent,
  //    research-expertise) + the spec-forge suite, so it travels too — the assembler wires sensei's skills
  //    from <dest>/skills/ (built-ins) via installNamedSkills.
  for (const sub of ["expertise", "hooks", "skills", "team", "install", "bin"]) {
    const s = path.join(gh, sub);
    if (!isDir(s)) {
      fail(`genesis home is missing ${sub}/ at ${gh} — is ${gh} a Genesis install?`);
    }
    copyTree(s, path.join(dest, sub));
  }

  // 1b. Stage the native `genesis-hook` binary into <dest>/bin so the enforcement hooks invoke it
  //     DIRECTLY (no Node in the hot path). The launcher (copied above into <dest>/bin) resolves it:
  //     GENESIS_HOOK_BIN for dev/CI, else download from the GitHub Release + SHA256-verify. Non-fatal +
  //     timeout-bounded: if staging fails, run.js + the agents fail-open, so the repo still bootstraps —
  //     the deterministic checks stay dormant until the binary is present. `<dest>/bin` is machine-local
  //     + regenerable (covered by the .gitignore).
  const binDir = path.join(dest, "bin");
  const launcher = path.join(binDir, "genesis-memory.js");
  const hookExe = process.platform === "win32" ? "genesis-hook.exe" : "genesis-hook";
  let hookBinStaged = false;
  try {
    fs.mkdirSync(binDir, { recursive: true });
    const r = spawnSync(process.execPath, [launcher, "--stage-hook", binDir], {
      encoding: "utf-8",
      timeout: 300000, // a cold cache may download the binary from the release
    });
    hookBinStaged = r.status === 0 && fs.existsSync(path.join(binDir, hookExe));
  } catch (e) {
    hookBinStaged = false;
  }

  // 2. Memory DB lives under .genesis/ (the launched server creates the file on first use). Never clobber
  //    an existing one. The DB is machine-local + regenerable; the PORTABLE, COMMITTED form of memory is the
  //    JSONL export at .genesis/memory/memory.jsonl (see the .gitignore in step 4). On a fresh clone the DB is
  //    absent but the JSONL is present, so the server rebuilds (re-embeds) the memory on first run.
  const memDb = path.join(dest, "memory.db");
  const memExport = path.join(dest, "memory", "memory.jsonl");

  // 3. Register the repo-local memory MCP server in <repo>/.mcp.json (merge; preserve other servers). The
  //    server is launched via the Node launcher staged in step 1 (.genesis/bin/genesis-memory.js), which
  //    downloads + SHA256-verifies the platform binary + model from the GitHub Release and caches them — NO
  //    local Rust build, and the launcher points the server at the cached model (no GENESIS_MODEL_DIR here).
  //    The repo-local DB + portable export keep memory per-repo and travelling with the repo across systems.
  const mcpPath = path.join(target, ".mcp.json");
  let mcp;
  try {
    mcp = JSON.parse(fs.readFileSync(mcpPath, { encoding: "utf-8" }));
  } catch (e) {
    mcp = {};
  }
  if (mcp === null || typeof mcp !== "object" || Array.isArray(mcp)) {
    mcp = {};
  }
  if (!mcp.mcpServers || typeof mcp.mcpServers !== "object") {
    mcp.mcpServers = {};
  }
  mcp.mcpServers["genesis-memory"] = {
    command: "node",
    args: [launcher],
    env: {
      GENESIS_MEMORY_DB: memDb,
      // The committed, cross-system-portable mirror. The server snapshots to this after every
      // store/consolidate and rebuilds from it when the DB is empty (fresh clone).
      GENESIS_MEMORY_EXPORT: memExport,
    },
  };
  fs.writeFileSync(mcpPath, JSON.stringify(mcp, null, 2) + "\n", { encoding: "utf-8" });

  // 4. Ensure the repo's .gitignore commits the agent BRAIN (expertise + hooks) and PORTABLE MEMORY
  //    (.genesis/memory/*.jsonl) while ignoring machine-local junk (the DB, caches, logs, the absolute-path
  //    .mcp.json). Without this, agents' memory + corrections never travel across systems.
  const gitignoreNote = mergeGitignore(target);

  // 5. Install sensei + method into <repo>/.claude/agents/, wired to the REPO-LEVEL .genesis/ home.
  const assembler = path.join(dest, "install", "assemble.js");
  const installed = [];
  for (const name of TEAM) {
    const src = path.join(dest, "team", name);
    const r = spawnSync(process.execPath, [assembler, src, name, target, dest], { encoding: "utf-8" });
    if (r.status !== 0) {
      fail(`assemble ${name} failed: ${((r.stderr || "").trim()) || ((r.stdout || "").trim())}`);
    }
    installed.push(JSON.parse((r.stdout || "").trim()).agent);
  }

  console.log(
    JSON.stringify(
      {
        bootstrapped: dest,
        target_repo: target,
        agents_installed: installed,
        agents_dir: path.join(target, ".claude", "agents"),
        hook_bin: hookBinStaged ? path.join(binDir, hookExe) : null,
        hook_bin_staged: hookBinStaged,
        memory_db: memDb,
        memory_db_preexisting: fs.existsSync(memDb),
        memory_export: memExport,
        memory_export_preexisting: fs.existsSync(memExport),
        gitignore: path.join(target, ".gitignore"),
        gitignore_warning: gitignoreNote,
        mcp_json: mcpPath,
        mcp_server: "node " + launcher,
        next: "Open Claude Code in the repo; talk to Sensei to build agents (expertise via the research-expertise skill).",
      },
      null,
      2
    )
  );
}

if (require.main === module) {
  main();
}

module.exports = { main, mergeGitignore, GITIGNORE_BLOCK, GITIGNORE_START, GITIGNORE_END };
